### PR TITLE
Task 16: Add Institute relations

### PR DESCRIPTION
**Pull request by Luke-kb**
## Description
This PR adds a has_many courses relation to the institute model, with course belongs_to institute. Course edit page adds institute dropdown selector. Course list page displays institute name. Institute list page displays all courses within an institute with each a link to the show page of that course.
In addition, RSpec github actions CI workflow and vscode devcontainer config is added.

Clarification from @jhuizy acting as 'Product Owner' determined that there is no requirement for a course to have an institute.
Therefore, the `null:false` attribute on `institute_id` in `courses` table was removed from the most recent migration. 

## Breakdown of Commits
Commit 1 : [27a1373](https://github.com/Luke-kb/mySMS/commit/27a1373f38664194148be8f033a1f0f0772f7414)
- devcontainer workflow config
- rspec testing github action setup
 
Commit 2 : [0b71a19](https://github.com/Luke-kb/mySMS/commit/0b71a19c3bf1dbaeb4d0a11fe36dd3143f917857)
- added institute foreign key to courses migration. updated models

Commit 3 : [1a5df6d](https://github.com/Luke-kb/mySMS/commit/1a5df6d719a902aff7c43eb0518032c84b176c72)
- added .env file

Commit 4 : [41aa41a](https://github.com/Luke-kb/mySMS/commit/41aa41af16f6bc616b068c9f12ba5b22f654bec9)
- dropdown input field for institutes for course edit page
- refactored indentation

Commit 5 : [87e55ab](https://github.com/Luke-kb/mySMS/commit/87e55abbfc5b883f32ee0b4e55065d0f57b960d9)
- modified factories and dev:populate

Commit 6 : [9461916](https://github.com/Luke-kb/mySMS/commit/94619166bceb69028d1f65d8d03095d9c08a06ff)
- small syntax changes
- refactor institute index

Commit 7 : [75b0d3e](https://github.com/Luke-kb/mySMS/commit/75b0d3e99e054b3e706e0e04b7d046bbed80e568)
- put github action in correct directory
- create navbar partial

Commit 8 : [f6df5a7](https://github.com/Luke-kb/mySMS/commit/f6df5a785356e4fae8dcd697eefdd63478a053b5)
- fixed indentation in rspec.yaml

Commit 9 : [2095b59](https://github.com/Luke-kb/mySMS/commit/2095b59f7bd892b7252784d6237dac3d2c3cfb17)
- renamed to rspec.yml
- added correct password to docker-compose.yml and database.yml and corrected mysql env variable name

Commit 10 : [e3235f3](https://github.com/Luke-kb/mySMS/commit/e3235f35fd526d8e54f645eb27eea7d678c57d01)
- syntax error

Commit 11 : [24dc164](https://github.com/Luke-kb/mySMS/commit/24dc164c2bab09e157f8ffc0126f2b6576d282b8)
- added status badge

Commit 12 : [0915b60](https://github.com/Luke-kb/mySMS/commit/0915b601a5fea8c702f88b3045595da3a1c60b8a)
- deleted commented code. 
- fixed N+1 queries

Commit 13 : [c21588c](https://github.com/Luke-kb/mySMS/commit/c21588c641f739d74f692822d76677ec3cdfb00e)
- added safe operator

Commit 14 : [87553bc](https://github.com/Luke-kb/mySMS/commit/c21588c641f739d74f692822d76677ec3cdfb00e)
- modified migration to be reversible

Commit 15 : [1b8c8e3](https://github.com/Luke-kb/mySMS/commit/1b8c8e3f22bd91ab1510fafa7b04dda8383ace76)
- removed null:false. courses can exist without an institute_id

Commit 16: [cd031de](https://github.com/Luke-kb/mySMS/commit/cd031de75af5cbe1f0fc7c2aaa78bd0efdb428b5)
- removed trailing .all 
- shorthand factory association attribute

Commit 17 : [c3adcd2](https://github.com/Luke-kb/mySMS/commit/c3adcd21efbf6f6749fe952bc2046fc338bfd64f)
- refactor of populate take task

Commit 18 : [4ea1253](https://github.com/Luke-kb/mySMS/commit/4ea1253ceedeb6dfd007dbc2e3b6082d69a66eda
- revised populate.rake refactor. removed rand integer assignment
- corrected shorthand factory association 
- added hard coded array of titles to be sampled on each factorybot create/build

#### TODO Checklist
* [x] setup CI github action workflow
* [x] setup vscode devcontainer workflow
* [x] add relations to institute and course models
* [x] add index and show dropdown menus and links
* [x] removed `null:false` from institute_id column in courses table 
* [x] refactor of populate.rake task

### CODE REVIEW

File: app/views/courses/index.html.haml [hunk 1]
```diff
@@ -20,6 +21,7 @@
       %td= course.start_date
       %td= course.end_date
       %td= course.allocation
+      %td= course.institute.name ? course.institute.name : "N/A"
```
Comment (Luke-kb) [inline app/views/courses/index.html.haml]: not sure about this...

File: app/views/courses/edit.html.haml [hunk 2]
```diff
@@ -3,25 +3,28 @@
...
       = f.label :allocation, "Allocation"
       = f.number_field :allocation
-    %div.form-group
+    .form-group
+      = f.label :institute, "Institute"
+      = f.select_tag :institute, options_from_collection_for_select(Institute.all, :id, :name)
+    .form-group
       = f.submit "Update", class: 'btn btn-success'
 
 = link_to 'Back', courses_path, class: 'btn btn-info'
```
Comment (josh-cooper) [inline app/views/courses/edit.html.haml:30]: 👍

File: app/views/courses/index.html.haml (hunk 1 above)
Comment (josh-cooper) [reply app/views/courses/index.html.haml]: Why have a conditional case for `name` but not the other attrs. Isn't institute `name` presence already validated?

If you're checking that the institute is valid you should probably safe navigate (`course&.institute.name`) or use `validates_associated` in the course model.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	// Uncomment the next line if you want to keep your containers running after VS Code shuts down.
+	// "shutdownAction": "none",
+
+	// Uncomment the next line to run commands after the container is created - for example installing curl.
+	// "postCreateCommand": "rails s",
+
+	// Uncomment to connect as a non-root user if you've added one. See https://aka.ms/vscode-remote/containers/non-root.
+	// "remoteUser": "vscode"
+}
```
Comment (josh-cooper) [inline .devcontainer/devcontainer.json:60]: Nice 😁

File: app/views/courses/show.html.haml [hunk 4]
```diff
@@ -19,6 +20,7 @@
             %td= @course.start_date
             %td= @course.end_date
             %td= @course.allocation
+            %td= @course.institute
```
Comment (josh-cooper) [inline app/views/courses/show.html.haml]: What does @course.institute output? I can't find a to_s. 
If it prints the institute name, why not use that above in the index page?

**Review by josh-cooper — COMMENTED** [review]
Do the github action run? I can't see the little green tick.

LGTM apart from some minor comments.

File: app/views/courses/show.html.haml (hunk 4 above)
Comment (Luke-kb) [reply app/views/courses/show.html.haml]: Ah, thanks. That should have been `@course.institute.name` - totally missed that

File: .github/workflows/rspec.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    runs-on: ubuntu-16.04
+    services:
+      mysql:
+        image: mysql:5.7
+        ports:
+          - 3306:3306
+        env:
+          MYSQL_USER: ${{ env.MYSQL_USER }}
+          MYSQL_ROOT_PASSWORD: ${{ env.MYSQL_ROOT_PASSWORD }}
+        # options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=10
```
Comment (jhuizy) [inline .github/workflows/rspec.yml]: can probably just remove the comments 👍

File: app/views/courses/index.html.haml [hunk 6]
```diff
@@ -20,6 +21,7 @@
       %td= course.start_date
       %td= course.end_date
       %td= course.allocation
+      %td= course&.institute.name
```
Comment (jhuizy) [inline app/views/courses/index.html.haml]: if course is nil then this will throw an error. If you never expect course to be nil you can remove the safe operator

```suggestion
      %td= course.institute.name
```

OR

```suggestion
      %td= course&.institute&.name
```

File: app/views/institutes/index.html.haml [hunk 7]
```diff
@@ -1,16 +1,23 @@
...
+    %th Courses
 
-    - @institutes.each do |institute|
-        %tr
-            %td= institute.name
-            %td= link_to "Show", institute_path(institute)
-            %td= link_to "Edit", edit_institute_path(institute)
+  - @institutes.each do |institute|
+    %tr
+      -# %td= institute.name
```
Comment (jhuizy) [inline app/views/institutes/index.html.haml]: commented code can be removed or fixed 👍

File: app/views/courses/index.html.haml (hunk 6 above)
Comment (jhuizy) [reply app/views/courses/index.html.haml]: Possible N+1 query here, I'm not sure how you're loading `@courses` but you'll need something like

```ruby
@courses = Course.includes(:institutes).all
```

in your controller

File: app/views/institutes/index.html.haml [hunk 8]
```diff
@@ -1,16 +1,23 @@
...
+      %td= link_to "#{institute.name}", institute_path(institute)
+      %td 
+        %ul.list-unstyled
+          %li= link_to "Edit", edit_institute_path(institute)
+      %td
+        %ul.list-unstyled
+          - institute.courses.each do |course|
```
Comment (jhuizy) [inline app/views/institutes/index.html.haml:17]: Another possible N+1

File: app/views/layouts/application.html.haml [hunk 9]
```diff
@@ -7,20 +7,9 @@
     %title= content_for?(:title) ? yield(:title) : "MySMS"
     = csrf_meta_tags
     = stylesheet_link_tag "application", :media => "all"
-    = javascript_include_tag "application"
+    -# = javascript_include_tag "application"
```
Comment (jhuizy) [inline app/views/layouts/application.html.haml]: why this change?

File: db/migrate/20210325004818_add_institute_reference_to_course_table.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class AddInstituteReferenceToCourseTable < ActiveRecord::Migration
+  def change
+    # add foreign key institute_id column
+    change_table :courses do |t|
+      t.references :institute, null: false
```
Comment (jhuizy) [inline db/migrate/20210325004818_add_institute_reference_to_course_table.rb]: Does this work? I don't think we can enforce `null` here since we have existing courses. We'd need a default value if that were the case. 

Also worth noting that this only creates a `institute_id` column, and not an actual foreign key. You will need to use an external gem such as [foreigner](https://github.com/matthuhiggins/foreigner) to create FKs, but I think that is outside the scope of these tasks. FWIW, plus has it's own implementation of creating and managing foreign keys. 

For this, I think it's enough that we just create the reference and allow nulls at the database level, and in our code we can enforce presence with an activerecord validation (although - do we even need to enforce that?)

**Review by jhuizy — CHANGES_REQUESTED** [review]
Couples of changes to look at, but otherwise nice work so far 👍

File: app/views/courses/index.html.haml (hunk 6 above)
Comment (Luke-kb) [reply app/views/courses/index.html.haml]: Thanks - didn't realise about the N+1

File: db/migrate/20210325004818_add_institute_reference_to_course_table.rb (hunk 10 above)
Comment (Luke-kb) [reply db/migrate/20210325004818_add_institute_reference_to_course_table.rb]: This code seemed to work fine. But I didn't know about needing a gem to create a foreign key! I'd say we probably don't need to enforce presence via AR validation.

File: app/views/layouts/application.html.haml (hunk 9 above)
Comment (Luke-kb) [reply app/views/layouts/application.html.haml]: oops! forgot to uncomment.

File: spec/factories/courses.rb [hunk 11]
```diff
@@ -1,10 +1,12 @@
...
   factory :course do
-    name { "#{Faker::Job.unique.field} 101" }
+    association :institute
+    name { "#{Faker::Educator.unique.course_name}" }
     description { (Faker::Lorem.paragraphs(number: 2, supplemental: true)).join(" ") }
     number_of_semesters { rand(1..4) }
     start_date { Faker::Date.between(from: 10.days.ago, to: Date.today) }
     end_date { Faker::Date.forward(days: rand(30..50)) }
     allocation { rand(20..40) }
+    institute_id { rand(1..10) }
```
Comment (jhuizy) [inline spec/factories/courses.rb]: No need for this as it's covered in the association field above

File: spec/factories/courses.rb [hunk 12]
```diff
@@ -1,10 +1,12 @@
 FactoryBot.define do
   factory :course do
-    name { "#{Faker::Job.unique.field} 101" }
+    association :institute
```
Comment (jhuizy) [inline spec/factories/courses.rb]: This is identical to just `institute`. What you have is fine but just if you wanted a more shorthand way of doing it in future.

```suggestion
    institute
```

File: app/views/courses/index.html.haml [hunk 13]
```diff
@@ -20,6 +21,7 @@
       %td= course.start_date
       %td= course.end_date
       %td= course.allocation
+      %td= course&.institute&.name
```
Comment (jhuizy) [inline app/views/courses/index.html.haml]: Looking at the other methods called, you won't need the safe operator for accessing institute, since `course` is proven not to be nil based on previous calls

```suggestion
      %td= course.institute&.name
```

File: app/controllers/courses_controller.rb [hunk 14]
```diff
@@ -2,7 +2,7 @@ class CoursesController < ApplicationController
   before_filter :load_course, except: [:index]
 
   def index
-    @courses = Course.order(:name).paginate(page: params[:page], per_page: 10)
+    @courses = Course.order(:name).paginate(page: params[:page], per_page: 5).includes(:institute).all
```
Comment (jhuizy) [inline app/controllers/courses_controller.rb]: The trailing `.all` is not needed, and as a preference, I like to put the `includes` at the beginning of the query

```suggestion
    @courses = Course.includes(:institute).order(:name).paginate(page: params[:page], per_page: 5)
```

File: app/controllers/institutes_controller.rb [hunk 15]
```diff
@@ -2,7 +2,7 @@ class InstitutesController < ApplicationController
   before_filter :load_institute, except: [:index]
 
   def index
-    @institutes = Institute.paginate(page: params[:page], per_page: 10)
+    @institutes = Institute.order(:name).paginate(page: params[:page], per_page: 10).includes(:courses).all
```
Comment (jhuizy) [inline app/controllers/institutes_controller.rb]: ditto to the other comment

**Review by jhuizy — CHANGES_REQUESTED** [review]
Nice work, just a few more things

File: db/migrate/20210325004818_add_institute_reference_to_course_table.rb (hunk 10 above)
Comment (jhuizy) [reply db/migrate/20210325004818_add_institute_reference_to_course_table.rb]: Just for future reference, if we enforce `null: false` at the database level, we must also do the same at the active record level. Otherwise, active record will think a record is `valid?` since it only runs the "AR" validations, and then an error will be thrown by mysql when it attempts to save the record.

File: spec/factories/courses.rb (hunk 11 above)
Comment (Luke-kb) [reply spec/factories/courses.rb]: Thanks - this cleared up some confusion I had around using the factory when populating the db vs testing.

File: spec/factories/courses.rb [hunk 16]
```diff
@@ -1,6 +1,7 @@
 FactoryBot.define do
   factory :course do
-    name { "#{Faker::Job.unique.field} 101" }
+    :institute
```
Comment (jhuizy) [inline spec/factories/courses.rb]: Drop the leading colon

```suggestion
    institute
```

File: lib/tasks/dev/populate.rake [hunk 17]
```diff
@@ -2,18 +2,23 @@ require 'faker'
...
     # Student table
-    FactoryBot.create_list(:student, 50, title_id: rand(1..6))  
+    20.times do
+      FactoryBot.create(:student, title_id: rand(1..6))
+    end
+    
     # Course table
-    FactoryBot.create_list(:course, 20)
+    20.times do 
+      FactoryBot.create(:course, institute_id: rand(1..10))
```
Comment (jhuizy) [inline lib/tasks/dev/populate.rake]: If you leave the institute field blank it will automatically create an institute for you

```suggestion
      FactoryBot.create(:course)
```

Comment (jhuizy) [reply lib/tasks/dev/populate.rake]: That would be preferred over assigning random integers to the relation field

**Review by jhuizy — CHANGES_REQUESTED** [review]
Almost there 😄

File: spec/factories/courses.rb (hunk 16 above)
Comment (Luke-kb) [reply spec/factories/courses.rb]: 🤦 🙏

File: lib/tasks/dev/populate.rake (hunk 17 above)
Comment (Luke-kb) [reply lib/tasks/dev/populate.rake]: Not sure if this is standard practice but atm I'm assigning the random institute_id so that it populates courses and students with different institutes and titles respectively. Is there a better way that you could recommend?

Comment (jhuizy) [reply lib/tasks/dev/populate.rake]: Yea the suggested change above will create a course that has a institute (because of our factorybot definition). Alternatively you could create a institute and pass that into a course

```ruby
instutute = FactoryBot.create(:institute)
course = FactoryBot.create(:course, institute: institute)
```

There should be no need to mess with passing in IDs here

**Review by jhuizy — APPROVED** [review]

**Comment by jhuizy** [comment]
Nice work
